<h1 align="center">Linear Regression</h1>
<h2 align="center">Mobile applications vs website</h2>
<h3 align="center">Rositsa Chankova</h3>

### Ecommerce company tries to find where to focus - mobile applications vs website

## 1. Imports:

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## 2. Get the Data:

In [ ]:
customers = pd.read_csv("Ecommerce Customers.csv")

In [4]:
customers.head()

,Email,Address,Avatar,Avg. Session Length,Time on App,Time on Website,Length of Membership,Yearly Amount Spent
0,mstephenson@fernandez.com,"835 Frank Tunnel\nWrightmouth, MI 82180-9605",Violet,34.497268,12.655651,39.577668,4.082621,587.951054
1,hduke@hotmail.com,"4547 Archer Common\nDiazchester, CA 06566-8576",DarkGreen,31.926272,11.109461,37.268959,2.664034,392.204933
2,pallen@yahoo.com,"24645 Valerie Unions Suite 582\nCobbborough, D...",Bisque,33.000915,11.330278,37.110597,4.104543,487.547505
3,riverarebecca@gmail.com,"1414 David Throughway\nPort Jason, OH 22070-1220",SaddleBrown,34.305557,13.717514,36.721283,3.120179,581.852344
4,mstephens@davidson-herman.com,"14023 Rodriguez Passage\nPort Jacobville, PR 3...",MediumAquaMarine,33.330673,12.795189,37.536653,4.446308,599.406092


In [5]:
customers.describe()

,Avg. Session Length,Time on App,Time on Website,Length of Membership,Yearly Amount Spent
count,500.000000,500.000000,500.000000,500.000000,500.000000
mean,33.053194,12.052488,37.060445,3.533462,499.314038
std,0.992563,0.994216,1.010489,0.999278,79.314782
min,29.532429,8.508152,33.913847,0.269901,256.670582
25%,32.341822,11.388153,36.349257,2.930450,445.038277
50%,33.082008,11.983231,37.069367,3.533975,498.887875
75%,33.711985,12.753850,37.716432,4.126502,549.313828
max,36.139662,15.126994,40.005182,6.922689,765.518462


In [6]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Email                 500 non-null    object 
 1   Address               500 non-null    object 
 2   Avatar                500 non-null    object 
 3   Avg. Session Length   500 non-null    float64
 4   Time on App           500 non-null    float64
 5   Time on Website       500 non-null    float64
 6   Length of Membership  500 non-null    float64
 7   Yearly Amount Spent   500 non-null    float64
dtypes: float64(5), object(3)
memory usage: 31.4+ KB


## 3. Exploratory Data Analysis:

### Create a joint plot to compare "Time on Website" with "Yearly Amount Spent"

In [7]:
sns.set_palette("GnBu_d")
sns.set_style('whitegrid')

**More time on site, more money spent.**

In [8]:
sns.jointplot(x='Time on Website',y='Yearly Amount Spent',data=customers)

### Not much correlation, so do the same but with the "Time on App" column instead

In [9]:
sns.jointplot(x='Time on App',y='Yearly Amount Spent',data=customers)

### There is a little bit more correlation

### With jointplot create a 2D hex bin plot comparing "Time on App" and "Length of Membership"

In [10]:
sns.jointplot(x='Time on App',y='Length of Membership',kind='hex',data=customers)

## Explore  types of relationships across the entire data set. 

In [11]:
sns.pairplot(customers)

### Create a linear model plot of  "Yearly Amount Spen" vs. "Length of Membership" since it looks to be the most correlated feature

In [12]:
sns.lmplot(x='Length of Membership',y='Yearly Amount Spent',data=customers)

## 4. Training and Testing Data

### 4.1. Split the data into training and testing sets

**Variable "X" equal to the numerical features of the customers and Variable "y" equal to the "Yearly Amount Spent" column.**

In [13]:
y = customers['Yearly Amount Spent']

In [14]:
X = customers[['Avg. Session Length', 'Time on App','Time on Website', 'Length of Membership']]

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

### 4.2. Training the Model

In [17]:
from sklearn.linear_model import LinearRegression

In [18]:
#Create an instance of a LinearRegression() model 
model = LinearRegression() 

In [19]:
#Train/fit model on the training data
model.fit(X_train,y_train) 

LinearRegression()

In [20]:
#Print out the coefficients of the model
print('Coefficients: \n', model.coef_) 

Coefficients: 
 [25.98154972 38.59015875  0.19040528 61.27909654]


### 4.3. Predicting Test Data

In [21]:
predictions = model.predict( X_test)

In [22]:
plt.scatter(y_test,predictions)
plt.xlabel('Y Test')
plt.ylabel('Predicted Y')

Text(27.406250000000007, 0.5, 'Predicted Y')

#### At first glance model performs very well.

## 5. Evaluate the performance of the model

In [23]:
from sklearn import metrics

print('MAE:', metrics.mean_absolute_error(y_test, predictions))  #Mean Absolute Error
print('MSE:', metrics.mean_squared_error(y_test, predictions)) # Mean Squared Error
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, predictions))) #Root Mean Squared Error

MAE: 7.228148653430815
MSE: 79.81305165097409
RMSE: 8.933815066978614


In [24]:
metrics.explained_variance_score(y_test, predictions) #Variance score (R^2) - how much variance model explains

0.9890771231889607

### Model performs very well - 98%

## Residuals

In [25]:
sns.displot((y_test-predictions),bins=50);

### Histogram of the residuals looks normally distributed

## 6.  Conclusions

**<font color=blue> The question -  where to focus the efforts - on mobile app or website development?  
Recreate the Dataframe to check all the coefficients</font>**

In [26]:
coeffecients = pd.DataFrame(model.coef_,X.columns)
coeffecients.columns = ['Coeffecient']
coeffecients

,Coeffecient
Avg. Session Length,25.981550
Time on App,38.590159
Time on Website,0.190405
Length of Membership,61.279097


### Interpreting the coefficients - holding all other features fixed:

### 1 unit increase in:

1. **Avg. Session Length** is associated with an **increase of 25.98 total dollars spent**.
2. **Time on App** is associated with an **increase of 38.59 total dollars spent**.
3. **Time on Website** is associated with an **increase of 0.19 total dollars spent**.
4. **Length of Membership** is associated with an **increase of 61.27 total dollars spent**.